controllers/gcpmachine_controller_test.go

```diff
@@ -48,4 +62,163 @@ var _ = Describe("GCPMachineReconciler", func() {
 			Expect(result.RequeueAfter).To(BeZero())
 		})
 	})
+
+	Context("reconcile non-deleted GCPMachines", func() {
+		var (
+			reconciler   *GCPMachineReconciler
+			ctx          context.Context
+			clusterScope *scope.ClusterScope
+			machineScope *scope.MachineScope
+			mockCtrl     *gomock.Controller
+			gcpService   *mocks.MockServiceInterface
+		)
+		BeforeEach(func() {
+			flag.Set("logtostderr", "false")
+			flag.Set("v", "2")
+			klog.SetOutput(GinkgoWriter)
+			var err error
+			bootstrapData := "data"
+
+			mockCtrl = gomock.NewController(GinkgoT())
+			gcpService = mocks.NewMockServiceInterface(mockCtrl)
+
+			reconciler = &GCPMachineReconciler{
+				Client:     k8sClient,
+				Log:        log.Log,
+				computeSvc: gcpService,
+			}
+
+			gcpMachine := &infrav1.GCPMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo",
+					Namespace: "default",
+				},
+				Spec: infrav1.GCPMachineSpec{
+					Zone:         "us-west4-a",
+					InstanceType: "n1-standard-1",
+				},
+			}
+
+			machine := &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "fooMachine",
+					Namespace: "default",
+				},
+				Spec: clusterv1.MachineSpec{
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
+						Kind:       "GCPMachine",
+						Name:       gcpMachine.Name,
+					},
+					Bootstrap: clusterv1.Bootstrap{
+						Data: &bootstrapData,
+					},
+				},
+			}
+			Expect(k8sClient.Create(ctx, gcpMachine)).Should(Succeed())
+
+			cluster := &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cluster",
+					Namespace: "default",
+				},
+				Status: clusterv1.ClusterStatus{
+					InfrastructureReady: true,
+				},
+			}
+
+			gcpCluster := &infrav1.GCPCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "gcpcluster",
+					Namespace: "default",
+				},
+				Spec: infrav1.GCPClusterSpec{
+					Project: "gcp-project",
+				},
+			}
+
+			// Create the cluster scope
+			clusterScope, err = scope.NewClusterScope(scope.ClusterScopeParams{
+				Client:     k8sClient,
+				Logger:     log.Log,
+				Cluster:    cluster,
+				GCPCluster: gcpCluster,
+			})
+			Expect(err).To(BeNil())
+
+			// Create the machine scope
+			machineScope, err = scope.NewMachineScope(scope.MachineScopeParams{
+				Logger:     log.Log,
+				Client:     k8sClient,
+				Cluster:    cluster,
+				Machine:    machine,
+				GCPCluster: gcpCluster,
+				GCPMachine: gcpMachine,
+			})
+			Expect(err).To(BeNil())
+		})
+
+		AfterEach(func() {
+			Expect(k8sClient.Delete(ctx, machineScope.GCPMachine)).Should(Succeed())
+		})
+
+		It("should return early if gcp machine is in error state", func() {
```

Comment:
Can we change the test title to better reflect what is being tested? Currently it doesn't seem like we are testing early return. Is there a way to assert that reconciliation hasn't happened? Maybe verifying if ProviderID isn't set. since that is a mandatory step in reconciliation. WDYT?

Reply:
I'm not sure I understand your point? returning early is just another way of saying reconciliation didn't happen and since there is an error message it also means reconciliation didn't happen